### PR TITLE
Fix vertical bars due to FILD/FIST memcpy trick

### PR DESCRIPTION
**Pull request by kklobe**
# Description

This is a fix for the vertical bars video corruption seen on a few games (Carmageddon, Motor Mash) and demos (Sunflower, Multikolor), due to a trick they use with FILD/FIST FPU instructions to do a fast memcpy, specifically with 64-bit integer values.

Obviously this needs a lot of testing - throw everything you can at it.

## Related issues

Several games and demos are mentioned in #1231 and #3125.

Ideas for this fix were inspired by https://github.com/joncampbell123/dosbox-x/issues/119.

# Manual testing

I tested the following:

- Intel Math CoProcessor Advanced Diagnostics to verify no new floating point issues were introduced.
- Quake SW 1.06 to verify non-memcpy-trick usage of FILD/FIST still works
- Carmageddon
- Motor Mash
- Sunflower
- Multikolor

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by johnnovak** [comment]
Great stuff @kklobe. 🎖️  I guess we'll need to just merge it after some initial testing and live with it during the RC period.

Can you write a more detailed comment explaining how this hack is used, how it manifests in vertical bars (usually black), and mention a few games/demos that need it? Otherwise, people will be scratching their heads in the future... it's really not obvious from the code why this is necessary.

File: include/fpu.h [hunk 1]
```diff
@@ -68,6 +68,9 @@ struct FPU_P_Reg {
 	uint32_t d2 = 0;
 };
 
+// Used for FILD/FIST 64-bit memcpy fix
+constexpr uint8_t TAG_MEMCPY = 4;
```
Comment (johnnovak) [inline include/fpu.h:72]: Minor: TagMemCpy or something.

File: src/fpu/fpu_instructions.h [hunk 2]
```diff
@@ -20,6 +20,8 @@
 #include "fpu.h"
 #endif
 
+static constexpr int64_t int53_min = -(1LL << 53);
+static constexpr int64_t int53_max = (1LL << 53) - 1;
```
Comment (johnnovak) [inline src/fpu/fpu_instructions.h:24]: Int53Min, etc.

File: src/fpu/fpu_instructions.h [hunk 3]
```diff
@@ -198,7 +200,18 @@ static void FPU_FLD_I64(PhysPt addr,Bitu store_to) {
 	FPU_Reg blah;
 	blah.l.lower = mem_readd(addr);
 	blah.l.upper = mem_readd(addr+4);
-	fpu.regs[store_to].d = static_cast<Real64>(blah.ll);
+
+	// Only assign values that fit in the 53-bit mantissa. If the
+	// value exceeds that, assume it's the FILD/FIST 64-bit memcpy trick
+	// and memcpy the value into the FPU register.
+	if (blah.ll <= int53_max && blah.ll >= int53_min) {
```
Comment (johnnovak) [inline src/fpu/fpu_instructions.h:207]: Can we take the opportunity to rename `blah` to something else? Please? 😅

File: src/fpu/fpu_instructions.h [hunk 4]
```diff
@@ -266,10 +279,19 @@ static void FPU_FST_I32(PhysPt addr) {
 	mem_writed(addr,(val < 2147483648.0 && val >= -2147483648.0)?static_cast<int32_t>(val):0x80000000);
 }
 
-static void FPU_FST_I64(PhysPt addr) {
-	double val = FROUND(fpu.regs[TOP].d);
+static void FPU_FST_I64(PhysPt addr)
+{
 	FPU_Reg blah;
-	blah.ll = (val < 9223372036854775808.0 && val >= -9223372036854775808.0)?static_cast<int64_t>(val):LONGTYPE(0x8000000000000000);
+	// If the value was loaded with the FILD/FIST 64-bit memcpy trick,
```
Comment (johnnovak) [inline src/fpu/fpu_instructions.h:285]: `blah`... 🤦🏻

**Comment by kklobe** [comment]
> Great stuff @kklobe. 🎖️ I guess we'll need to just merge it after some initial testing and live with it during the RC period.
> 
> Can you write a more detailed comment explaining how this hack is used, how it manifests in vertical bars (usually black), and mention a few games/demos that need it? Otherwise, people will be scratching their heads in the future... it's really not obvious from the code why this is necessary.

Absolutely, I plan on adding more context, I just wanted to get this thing up for testing ASAP.

**Comment by johnnovak** [comment]
Tried a few demos and apps, and this works splendidly. Bear in mind that very very few DOS games and demos use the FPU, so 99% games & demos won't exercise the FPU paths _at all_. Also bear in mind that this trick only made sense in the Pentium era, so from 1995 onwards or so. You just won't see it in any earlier programs. Just saying before someone starts regression testing all pre-1990 CGA/EGA games with this fix in place 😎 

Otherwise, I'd say let's just merge it and see how it goes. All my testing has been positive so far! Outstanding work @kklobe 🚀

**Comment by Grounded0** [comment]
@johnnovak 

You're not the only one mixing these two up. At least now the PRs are correctly linked to the correct issues.

**Review by Grounded0 — APPROVED** [review]

**Review by johnnovak — APPROVED** [review]

**Comment by johnnovak** [comment]
Is there anything outstanding on this @kklobe? If not, let's press the Big Green Button 😎

**Comment by kklobe** [comment]
> Is there anything outstanding on this @kklobe? If not, let's press the Big Green Button 😎

Actually if you could give me just a bit more time - I have an idea for a little more robust fix when the value to be copied is outside the 53-bit integer range, shouldn't take long.

**Comment by kklobe** [comment]
> > Is there anything outstanding on this @kklobe? If not, let's press the Big Green Button 😎
> 
> Actually if you could give me just a bit more time - I have an idea for a little more robust fix when the value to be copied is outside the 53-bit integer range, shouldn't take long.

OK @johnnovak, I improved the out-of-range logic and fixed a couple other bugs. I ran a bunch of tests, but maybe give it another quick whirl on your preferred vertical bar apps.

And, if you were implying I should merge - I don't have (and don't want) "press the green button" permissions :grin:

File: src/fpu/fpu_instructions.h [hunk 5]
```diff
@@ -34,6 +34,10 @@ static void FPU_FINIT(void) {
 	fpu.tags[6] = TAG_Empty;
 	fpu.tags[7] = TAG_Empty;
 	fpu.tags[8] = TAG_Valid; // is only used by us
+
+	for (auto& use_regs_memcpy : fpu.use_regs_memcpy) {
+		use_regs_memcpy = false;
+	}
```
[the remark this answers is not in the document]
Comment (kklobe) [reply src/fpu/fpu_instructions.h:40]: That doesn't compile, the loop shouldn't really be a big deal for a one-time init.

File: src/fpu/fpu_instructions.h (hunk 4 above)
Comment (kcgen) [reply src/fpu/fpu_instructions.h:294]: `blah`, being a potential placeholder world in English, was recommended above less flexible interjectives like `argh`, `ugh`, `grrr`, and `doh` 🤦

**Comment by kklobe** [comment]
this raises a philosophical question: since integers between INT53_MAX and INT64_MAX likely will not be the same after storing and loading, should we stick with the old behavior, which returns a slightly inaccurate integer, or should we return a NaN, as I was doing? Is it better to be inaccurate but close or have the app in question barf on a NaN?

For example, here's a quick test:

```cpp
#include <cstdio>

#define INT53_MAX ((1LL << 53) - 1)
#define INT53_MIN -(1LL << 53)

int main(int argc, char *argv[]) {
	for (int64_t i = INT53_MAX - 10; i<INT53_MAX + 10;i++) {
		double dbl = static_cast<double>(i);
		const int64_t casted = static_cast<int64_t>(dbl);
		printf("%lld, %f, %lld, %s\n", i, dbl, casted, i == casted ? "PASS" : "FAIL");
	}
}
```

and the output:

```log
9007199254740981, 9007199254740981.000000, 9007199254740981, PASS
9007199254740982, 9007199254740982.000000, 9007199254740982, PASS
9007199254740983, 9007199254740983.000000, 9007199254740983, PASS
9007199254740984, 9007199254740984.000000, 9007199254740984, PASS
9007199254740985, 9007199254740985.000000, 9007199254740985, PASS
9007199254740986, 9007199254740986.000000, 9007199254740986, PASS
9007199254740987, 9007199254740987.000000, 9007199254740987, PASS
9007199254740988, 9007199254740988.000000, 9007199254740988, PASS
9007199254740989, 9007199254740989.000000, 9007199254740989, PASS
9007199254740990, 9007199254740990.000000, 9007199254740990, PASS
9007199254740991, 9007199254740991.000000, 9007199254740991, PASS
9007199254740992, 9007199254740992.000000, 9007199254740992, PASS
9007199254740993, 9007199254740992.000000, 9007199254740992, FAIL
9007199254740994, 9007199254740994.000000, 9007199254740994, PASS
9007199254740995, 9007199254740996.000000, 9007199254740996, FAIL
9007199254740996, 9007199254740996.000000, 9007199254740996, PASS
9007199254740997, 9007199254740996.000000, 9007199254740996, FAIL
9007199254740998, 9007199254740998.000000, 9007199254740998, PASS
9007199254740999, 9007199254741000.000000, 9007199254741000, FAIL
9007199254741000, 9007199254741000.000000, 9007199254741000, PASS
```

EDIT: nevermind for now, we'll leave the logic the way it's always been and address any bugs resulting from legitimate use of large numbers for non-memcpy-trick scenarios when (and if) they arise.

**Comment by kcgen** [comment]
> this raises a philosophical question: since integers between INT53_MAX and INT64_MAX likely will not be the same after storing and loading, should we stick with the old behavior, which returns a slightly inaccurate integer, or should we return a NaN, as I was doing? Is it better to be inaccurate but close or have the app in question barf on a NaN?
> 
> For example, here's a quick test:
> 
> ```c++
> #include <cstdio>
> 
> #define INT53_MAX ((1LL << 53) - 1)
> #define INT53_MIN -(1LL << 53)
> 
> int main(int argc, char *argv[]) {
> 	for (int64_t i = INT53_MAX - 10; i<INT53_MAX + 10;i++) {
> 		double dbl = static_cast<double>(i);
> 		const int64_t casted = static_cast<int64_t>(dbl);
> 		printf("%lld, %f, %lld, %s\n", i, dbl, casted, i == casted ? "PASS" : "FAIL");
> 	}
> }
> ```
> 
> and the output:
> 
> ```
> 9007199254740981, 9007199254740981.000000, 9007199254740981, PASS
> 9007199254740982, 9007199254740982.000000, 9007199254740982, PASS
> 9007199254740983, 9007199254740983.000000, 9007199254740983, PASS
> 9007199254740984, 9007199254740984.000000, 9007199254740984, PASS
> 9007199254740985, 9007199254740985.000000, 9007199254740985, PASS
> 9007199254740986, 9007199254740986.000000, 9007199254740986, PASS
> 9007199254740987, 9007199254740987.000000, 9007199254740987, PASS
> 9007199254740988, 9007199254740988.000000, 9007199254740988, PASS
> 9007199254740989, 9007199254740989.000000, 9007199254740989, PASS
> 9007199254740990, 9007199254740990.000000, 9007199254740990, PASS
> 9007199254740991, 9007199254740991.000000, 9007199254740991, PASS
> 9007199254740992, 9007199254740992.000000, 9007199254740992, PASS
> 9007199254740993, 9007199254740992.000000, 9007199254740992, FAIL
> 9007199254740994, 9007199254740994.000000, 9007199254740994, PASS
> 9007199254740995, 9007199254740996.000000, 9007199254740996, FAIL
> 9007199254740996, 9007199254740996.000000, 9007199254740996, PASS
> 9007199254740997, 9007199254740996.000000, 9007199254740996, FAIL
> 9007199254740998, 9007199254740998.000000, 9007199254740998, PASS
> 9007199254740999, 9007199254741000.000000, 9007199254741000, FAIL
> 9007199254741000, 9007199254741000.000000, 9007199254741000, PASS
> ```

When the int64 value goes beyond 53-bits in size, we could "spillover" those lost bits of precision into a new `uint16_t` variable and then  we  recover it on the backside when casting the double back to an `int64_t`.

Something like this:

```c++
#include <cassert>
#include <cmath>
#include <cstdint>
#include <cstdio>

#define INT53_MAX ((1LL << 53) - 1)
#define INT53_MIN -(1LL << 53)
#define SPILLBITS 12

using double_spillover_t = std::pair<double, uint16_t>;

static double_spillover_t
maybe_spillover(const int64_t value)
{
  const auto in_range = (std::abs(value) <= INT53_MAX);
  if (in_range) {
    return { static_cast<double>(value), 0 };
  }
  return { static_cast<double>(value >> SPILLBITS),
           static_cast<uint16_t>(value & 0b1111'1111'1111) };
}

static constexpr int64_t
to_int64_from_spillover(const double_spillover_t ds)
{
  return ds.second == 0
           ? static_cast<int64_t>(ds.first)
           : ((static_cast<int64_t>(ds.first) << SPILLBITS) | ds.second);
}

int
main(int argc, char* argv[])
{
  for (int64_t i = INT53_MAX + 1; i < INT64_MAX; i *= 10) {
    const auto ds = maybe_spillover(i);
    const int64_t casted = to_int64_from_spillover(ds);
    if (i != casted) {
      printf("FAIL: %lld, %f, %lld", i, ds.first, casted);
    }
  }
  return 0;
}
```

**Comment by johnnovak** [comment]
> When the int64 value goes beyond 53-bits in size, we could "spillover" those lost bits of precision (when casting to `double`) into a new `uint16_t` variable and then we recover the bits on backside when casting the `double` back to an `int64_t`.

I think that would be the safest—the question is how "hot" this path would be and the performance impact compared to returning slighlty inaccurate results. I don't think returning NaN is a good idea; at the very least we want the old behaviour I think.

**Comment by kklobe** [comment]
> > When the int64 value goes beyond 53-bits in size, we could "spillover" those lost bits of precision (when casting to `double`) into a new `uint16_t` variable and then we recover the bits on backside when casting the `double` back to an `int64_t`.
> 
> I think that would be the safest—the question is how "hot" this path would be and the performance impact compared to returning slighlty inaccurate results. I don't think returning NaN is a good idea; at the very least we want the old behaviour I think.

I'm not even sure this would help the scenario I'm thinking of: when someone isn't caliing FPU_FST_I64 as part of the memcpy FILD/FIST pair. Let's say they do a direct load of the FP register that ends up in fpu.regs[TOP] (the code in the `else` block below), and it's > int53_max but <= int64_max, but they really do want that large value. We've already lost the precision at that point.

```cpp
static void FPU_FST_I64(PhysPt addr)
{
	FPU_Reg temp_reg;
	// If the value was loaded with the FILD/FIST 64-bit memcpy trick,
	// read the 64-bit integer value instead.
	if (fpu.use_regs_memcpy[TOP]) {
		temp_reg.ll = fpu.regs_memcpy[TOP];
	} else {
		double val = FROUND(fpu.regs[TOP].d);
		temp_reg.ll = (val <= static_cast<double>(INT64_MAX) &&
		               val >= static_cast<double>(INT64_MIN))
		                    ? static_cast<int64_t>(val)
		                    : LONGTYPE(0x8000000000000000);
	}

	mem_writed(addr, temp_reg.l.lower);
	mem_writed(addr + 4, temp_reg.l.upper);
}

```
One thought I had was that we could set the floating point exception for Invalid Operation and/or Loss of Precision. From the Intel CPU ref for `FIST`:

```log
Floating-Point Exceptions
#IS Stack underflow occurred.
#IA Converted value is too large for the destination format.
Source operand is an SNaN, QNaN, ±∞, or unsupported format. #P Value cannot be represented exactly in destination format.
```

But yes, let's keep the old inaccurate behavior for now, until we have a test case in a real app 😁

**Comment by johnnovak** [comment]
> But yes, let's keep the old inaccurate behavior for now, until we have a test case in a real app 😁

Not broken, don't fix it 😎 Sounds good to me.

**Comment by kklobe** [comment]
Sorry guys - added quadword memory functions and used them in the FPU load/store functions, a bit more to review.

**Comment by kklobe** [comment]
OK, fixed most of the embarrassing mistakes, ready for scrutiny.

**Comment by kklobe** [comment]
As far as the quadword functions go, I saw that the `FPU_FST_F64` and `FPU_FLD_F64` are very heavily called, both in the games that need the memcpy hack, as well as those that don't.

Quake calls them several hundred times per millisecond:

```log
2023-11-27 22:31:14.386 | FPU_FST_F64
2023-11-27 22:31:14.386 | FPU_FST_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
2023-11-27 22:31:14.386 | FPU_FST_F64
2023-11-27 22:31:14.386 | FPU_FLD_F64
```

So I figured cutting the memory-related calls in half was a good thing, and should benefit many more games than just the memcpy-hack ones.

**Comment by kklobe** [comment]
setting this back to DRAFT since I rebased on @johnnovak's aspect PR.

**Comment by johnnovak** [comment]
Are we done? Merge time @kklobe ?

**Comment by kklobe** [comment]
> Are we done? Merge time @kklobe ?

No further changes from me for now, let's do it.

**Pull request merged**